opts/hugetlb.go

```diff
@@ -0,0 +1,113 @@
+package opts
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/docker/docker/api/types/container"
+	"github.com/docker/go-units"
+)
+
+// ValidatorHugetlbType defines a validator function that returns a validated struct and/or an error.
+type ValidatorHugetlbType func(val string) (container.Hugetlb, error)
+
+// ValidateHugetlb validates that the specified string has a valid hugetlb format.
+func ValidateHugetlb(htlb string) (container.Hugetlb, error) {
+	var size, limit string
+	var hugetlb container.Hugetlb
+
+	ss := strings.SplitN(htlb, ":", 2)
+	if len(ss) == 2 {
+		if ss[0] == "" {
+			size = ""
+		} else {
+			size = formatHugepageSize(ss[0])
+		}
+		limit = ss[1]
+	} else {
+		size = ""
+		limit = ss[0]
+	}
+
+	ilimit, err := units.RAMInBytes(limit)
+	if err != nil {
+		return hugetlb, fmt.Errorf("Invalid hugetlb limit:%s", limit)
+	}
+	ulimit := uint64(ilimit)
+	hugetlb = container.Hugetlb{
+		PageSize: size,
+		Limit:    ulimit,
+	}
+	return hugetlb, nil
+}
+
+// HugetlbOpt defines a map of Hugetlbs
+type HugetlbOpt struct {
+	values    []container.Hugetlb
+	validator ValidatorHugetlbType
+}
+
+// NewHugetlbOpt creates a new HugetlbOpt
+func NewHugetlbOpt(validator ValidatorHugetlbType) HugetlbOpt {
+	values := []container.Hugetlb{}
+	return HugetlbOpt{
+		values:    values,
+		validator: validator,
+	}
+}
+
+// Set validates a Hugetlb and sets its name as a key in HugetlbOpt
+func (opt *HugetlbOpt) Set(val string) error {
+	var value container.Hugetlb
+	if opt.validator != nil {
+		v, err := opt.validator(val)
+		if err != nil {
+			return err
+		}
+		value = v
+	}
+	(opt.values) = append((opt.values), value)
+
+	return nil
+}
+
+// String returns HugetlbOpt values as a string.
+func (opt *HugetlbOpt) String() string {
+	var out []string
+	for _, v := range opt.values {
+		out = append(out, fmt.Sprintf("%v", v))
+	}
+
+	return fmt.Sprintf("%v", out)
+}
+
+// GetAll returns a slice of pointers to Hugetlbs.
+func (opt *HugetlbOpt) GetAll() []container.Hugetlb {
+	var hugetlbs []container.Hugetlb
+	for _, v := range opt.values {
+		hugetlbs = append(hugetlbs, v)
+	}
+
+	return hugetlbs
+}
+
+// Type returns the option type
+func (opt *HugetlbOpt) Type() string {
+	return "hugetlb"
+}
+
+func formatHugepageSize(s string) string {
+	// make sure size hase suffix "B"
+	if (!strings.HasSuffix(s, "b")) &&
+		(!strings.HasSuffix(s, "B")) {
+		s = s + "B"
+	}
+
+	// make sure size get all 'b/k/m/g' replaced with "B/K/M/G"
+	s = strings.Replace(s, "b", "B", -1)
+	s = strings.Replace(s, "k", "K", -1)
+	s = strings.Replace(s, "m", "M", -1)
+	s = strings.Replace(s, "g", "G", -1)
```

Comment:
Why is this function necessary?

I think we should only support the formats that we support everywhere else from `units.RAMInBytes()`. Is there some reason we can't?

Looking at the code some more I'm guessing this is because the linux API excepts these values?

I'd like to propose this instead:
* parse both size using `units.RAMInBytes()` here in `opts` package
* have the API accept `int64` instead of string
* in `daemon.getHugetlbResources()` convert to the form expected by that API

I think that will remove all the duplicate validation, parsing and formatting. It will make the values consistent with other fields, and it will make the API consistent with newer API fields.